streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -1054,9 +1101,21 @@ private void suspendTasksAndState()  {
 
         final AtomicReference<RuntimeException> firstException = new AtomicReference<>(null);
 
-        for (final AbstractTask task : activeAndStandbytasks()) {
-            try {
+        firstException.compareAndSet(null, performOnStreamTasks(new StreamTaskAction() {
+            @Override
+            public String name() {
+                return "suspend";
+            }
+
+            @Override
+            public void apply(final StreamTask task) {
                 task.suspend();
+            }
+        }));
+
+        for (final StandbyTask task : standbyTasks.values()) {
```

Comment:
Yes, we should call `suspend` here -- but we should keep both loops and handle `StreamTask` and `StandbyTask` separately. As `StreamTask` with EOS enabled could be a zombie task (cannot happen for `StandbyTask`) and we want to remove zombie tasks from `activeTask`, the code is cleaner if we don't do a single loop (otherwise we need to do all kind of casts (and/or `instanceof`) and the iterator pattern that we use in `performOnStreamTasks` would also not be possible). Let me know if you disagree.